e2etester/runtests.go

```diff
@@ -0,0 +1,594 @@
+		test.w.Zone = t.Zone
+		test.w.OAuthPath = t.OAuthPath
+		test.w.ComputeEndpoint = t.ComputeEndpoint
+		test.w.DisableGCSLogging()
+		test.w.DisableCloudLogging()
+		test.w.DisableStdoutLogging()
+		test.w.Logger = test.logger
+	}
+
+	return &t, nil
+}
+
+func checkError(errors chan error) {
+	select {
+	case err := <-errors:
+		fmt.Fprintln(os.Stderr, "\n[TestRunner] Errors in one or more test cases:")
+		fmt.Fprintln(os.Stderr, "\n - ", err)
+		for {
+			select {
+			case err := <-errors:
+				fmt.Fprintln(os.Stderr, "\n - ", err)
+				continue
+			default:
+				fmt.Fprintln(os.Stderr, "\n[TestRunner] Exiting with exit code 1")
+				os.Exit(1)
+			}
+		}
+	default:
+		return
+	}
+}
+
+type junitTestSuite struct {
+	mx sync.Mutex
+
+	XMLName  xml.Name `xml:"testsuite"`
+	Name     string   `xml:"name,attr"`
+	Tests    int      `xml:"tests,attr"`
+	Failures int      `xml:"failures,attr"`
+	Errors   int      `xml:"errors,attr"`
+	Disabled int      `xml:"disabled,attr"`
+	Skipped  int      `xml:"skipped,attr"`
+	Time     float64  `xml:"time,attr"`
+
+	TestCase []*junitTestCase `xml:"testcase"`
+}
+
+type junitTestCase struct {
+	Classname string        `xml:"classname,attr"`
+	ID        string        `xml:"id,attr"`
+	Name      string        `xml:"name,attr"`
+	Time      float64       `xml:"time,attr"`
+	Skipped   *junitSkipped `xml:"skipped,omitempty"`
+	Failure   *junitFailure `xml:"failure,omitempty"`
+	SystemOut string        `xml:"system-out,omitempty"`
+}
+
+type junitSkipped struct {
+	Message string `xml:"message,attr"`
+}
+
+type junitFailure struct {
+	FailMessage string `xml:",chardata"`
+	FailType    string `xml:"type,attr"`
+}
+
+type test struct {
+	name     string
+	testCase *TestCase
+}
+
+func getCommonInstanceMetadata(client daisyCompute.Client, project string) (*compute.Metadata, error) {
+	proj, err := client.GetProject(project)
+	if err != nil {
+		return nil, fmt.Errorf("error getting project: %v", err)
+	}
+
+	return proj.CommonInstanceMetadata, nil
+}
+
+func delItem(items []*compute.MetadataItems, i int) []*compute.MetadataItems {
+	// Delete the element.
+	// https://github.com/golang/go/wiki/SliceTricks
+	copy(items[i:], items[i+1:])
+	items[len(items)-1] = nil
+	return items[:len(items)-1]
+}
+
+func isExpired(val string) bool {
+	t, err := time.Parse(timeFormat, val)
+	if err != nil {
+		return false
+	}
+	return time.Now().After(t)
+}
+
+const (
+	writeLock      = "TestWriteLock-"
+	readLock       = "TestReadLock-"
+	defaultTimeout = 2 * time.Hour
+)
+
+func waitLock(client daisyCompute.Client, project string, prefix ...string) (*compute.Metadata, error) {
+	var md *compute.Metadata
+	var err error
+Loop:
+	for {
+		md, err = getCommonInstanceMetadata(client, project)
+		if err != nil {
+			return nil, err
+		}
+
+		for i, mdi := range md.Items {
+			if mdi != nil {
+				for _, p := range prefix {
+					if strings.HasPrefix(mdi.Key, p) {
+						if isExpired(*mdi.Value) {
+							md.Items = delItem(md.Items, i)
```

Comment:
md is not thread-safe. Per my understanding to this PR, access to md is not isolated, right?

Reply:
this code was copied from daisy test runner. this locking function will happen during workflow execution, which happens strictly after any user-provided code has run. moreover, we are not supporting multithreaded uses of this code.